### PR TITLE
refactor preprocessing, one more time 

### PR DESCRIPTION
**Pull request by maxulysse**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by maxulysse** [comment]
reverted gatk4 spark containers to a version that was actually working

**Comment by FriederikeHanssen** [comment]
> reverted gatk4 spark containers to a version that was actually working

😢  I don't get why it doesn't work with the new one. Doesn't look like the base containers is changing. but great that you fixed it 🚀

**Comment by maxulysse** [comment]
> > reverted gatk4 spark containers to a version that was actually working
> 
> cry I don't get why it doesn't work with the new one. Doesn't look like the base containers is changing. but great that you fixed it rocket

I'm as confused as you are

File: .github/workflows/awsfulltest.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ name: nf-core AWS full size tests
 # It can be additionally triggered manually with GitHub actions workflow dispatch button.
 # It runs the -profile 'test_full' on AWS batch
 
-on:
+on: # yamllint disable-line rule:truthy
```
Comment (FriederikeHanssen) [inline .github/workflows/awsfulltest.yml:6]: is this from tools@dev?

File: conf/modules.config [hunk 2]
```diff
@@ -244,14 +244,12 @@ process {
     withName: "BWA.*MEM" {
         ext.args         = { meta.status == 1 ? '-K 100000000 -M -B 3' : '-K 100000000 -M' }
         ext.prefix       = { params.split_fastq > 1 ?  "${meta.id}".concat('.').concat(reads.get(0).name.findAll(/part_([0-9]+)?/).last()) : "" }
+        publishDir       = [
+            enabled: false
```
Comment (FriederikeHanssen) [inline conf/modules.config:248]: Isn't this the default? I think we discuss this in one of the previous PR'S?

File: conf/modules.config [hunk 3]
```diff
@@ -244,14 +244,12 @@ process {
...
+        ]
     }
 
-    withName: 'MERGE_MAPPING' {
-        ext.when         = { params.save_bam_mapped || (params.skip_tools && params.skip_tools.contains('markduplicates')) }
-    }
-
-    withName: 'INDEX_MAPPING' {
-        ext.when         = { params.save_bam_mapped || (params.skip_tools && params.skip_tools.contains('markduplicates')) }
+    withName: 'INDEX_MERGE_BAM' {
```
Comment (FriederikeHanssen) [inline conf/modules.config:252]: why did you remove the `when` here?

File: .github/workflows/awsfulltest.yml (hunk 1 above)
Comment (maxulysse) [reply .github/workflows/awsfulltest.yml:6]: No, that was me trying stuff out

File: conf/modules.config (hunk 2 above)
Comment (maxulysse) [reply conf/modules.config:248]: Wasn't working properly actually

File: .github/workflows/awsfulltest.yml (hunk 1 above)
Comment (maxulysse) [reply .github/workflows/awsfulltest.yml:6]: but I think actually I can remove all that out since I added the truthly thingy in the .yamllint

File: conf/modules.config (hunk 3 above)
Comment (maxulysse) [reply conf/modules.config:252]: No need for it anymore

File: conf/modules.config (hunk 2 above)
Comment (FriederikeHanssen) [reply conf/modules.config:248]: oh ok, so should be added back then to the other modules? Or is only BWA not working?

File: conf/modules.config (hunk 3 above)
Comment (maxulysse) [reply conf/modules.config:252]: I'll check again all `when` once done with the refactoring just to be sure all is as it should be

File: conf/modules.config (hunk 2 above)
Comment (maxulysse) [reply conf/modules.config:248]: no idea, I'll need to do some more proper tests with that

File: conf/modules.config (hunk 3 above)
Comment (FriederikeHanssen) [reply conf/modules.config:252]: ok Isee. Out of curiosity, why use an `if` statement in this case but for everything else `when`? Don't get me wrong, i still find `if` a lot easier to read, just want to learn more about this new syntax :)

Comment (maxulysse) [reply conf/modules.config:252]: Trying to find something nice to look at, easy to read, compact and easy enough to understand, so mainly trying things out :-D

Comment (FriederikeHanssen) [reply conf/modules.config:252]: ha, I suppose there will be many cycles until everyone is completely happy

File: conf/modules.config (hunk 2 above)
Comment (FriederikeHanssen) [reply conf/modules.config:260]: Is this solved?

**Review by FriederikeHanssen — APPROVED** [review]

File: conf/modules.config (hunk 2 above)
Comment (maxulysse) [reply conf/modules.config:260]: still not, I have no idea why I have the bam if I don't have this enabled: false here
